src/main/java/com/powsybl/openloadflow/sensi/DcSensitivityAnalysis.java

```diff
@@ -435,14 +432,79 @@ public DcSensitivityAnalysis(MatrixFactory matrixFactory) {
         super(matrixFactory);
     }
 
-    protected void setFunctionReferenceOnFactors(List<LfNetwork> lfNetworks, List<LfSensitivityFactor> factors,
-                                                 LoadFlowParameters lfParameters, OpenLoadFlowParameters lfParametersExt) {
-        DcLoadFlowParameters dcLoadFlowParameters = new DcLoadFlowParameters(lfParametersExt.getSlackBusSelector(), matrixFactory,
-                true, lfParametersExt.isDcUseTransformerRatio(), lfParameters.isDistributedSlack(),  lfParameters.getBalanceType());
-        DcLoadFlowResult dcLoadFlowResult = new DcLoadFlowEngine(lfNetworks, dcLoadFlowParameters).run();
+    protected DenseMatrix setReferenceActivePowerFlows(LfNetwork network, EquationSystem equationSystem, LUDecomposition lu, List<LfSensitivityFactor> factors,
+                                                       LoadFlowParameters lfParameters, List<ParticipatingElement> participatingElements, GraphDecrementalConnectivity<LfBus> connectivity) {
+
+        double[] x = equationSystem.createStateVector(new UniformValueVoltageInitializer());
+        Map<LfBus, BusState> busStates = new HashMap<>();
+        if (lfParameters.isDistributedSlack()) {
+            double mismatch;
+            if (connectivity != null) {
+                mismatch = network.getActivePowerMismatchInMainComponent(connectivity);
+            } else {
+                mismatch = network.getActivePowerMismatch();
+            }
+            busStates = BusState.createBusStates(participatingElements.stream()
+                    .map(ParticipatingElement::getLfBus)
+                    .collect(Collectors.toSet()));
+            int iteration = 0;
+            ActivePowerDistribution.Step step = ActivePowerDistribution.getStep(lfParameters.getBalanceType(), false);
+            while (!participatingElements.isEmpty()
+                    && Math.abs(mismatch) > ActivePowerDistribution.P_RESIDUE_EPS) {
+                mismatch -= step.run(participatingElements, iteration, mismatch);
+
+                iteration++;
+            }
+        }
+
+        equationSystem.updateEquations(x);
+
+        double[] dx = equationSystem.createTargetVector();
+
+        if (connectivity != null) {
+            // set buses injections and transformers to 0 outside the main connected component
+            int mainComponentNumber = connectivity.getComponentNumber(network.getSlackBus());
+            List<Integer> columnsToSetToZero = network.getBuses().stream()
+                    .filter(lfBus -> connectivity.getComponentNumber(lfBus) != mainComponentNumber)
```

Comment:
To get the buses which are not in the main component you should use directly `connectivity.getSmallComponents()`

Reply:
Unfortunately, I think getSmallComponents() returns the smallest components in size, but it has nothing to do with the slack bus (the slack could be in a small component). What I want here are the buses that are not in the same component than the slack bus.

Reply:
Indeed, but couldn't we make the assumption that the slack bus is in the biggest component in size? I thought you could, but well I might be mistaken - calling @annetill and @geofjamg to the rescue!